### DISCUSSION TITLE
npm warning when running rake assets:precompile

### DISCUSSION THREAD

**Discussion by jameshibbard** [ElMassimo/vite_ruby#221] [Q&A] [answered]
Hi,

Thank you for this great project.

This isn't a bug, so much as a question I couldn't find an answer to.

- [x] I have tried upgrading by running `bundle update vite_ruby`.
- [x] I have read the __[troubleshooting section]__ before opening an issue.

### Description 📖

When deploying with Capistrano, I see the following warning printed to the console:

```
01 npm
01  WARN config
01  production Use `--omit=dev` instead.
``` 
As far as I can tell, this warning seems to be caused by the `vite:install_dependencies` rake task.

### Reproduction 🐞

```
NODE_ENV=production rake assets:precompile --trace  
```

Output:

```
** Invoke assets:precompile (first_time)
** Invoke assets:environment (first_time)
** Execute assets:environment
** Invoke environment (first_time)
** Execute environment
** Execute assets:precompile
** Invoke vite:install_dependencies (first_time)
** Execute vite:install_dependencies
npm WARN config production Use `--omit=dev` instead.
npm WARN config production Use `--omit=dev` instead.
npm WARN config production Use `--omit=dev` instead.

added 275 packages, and audited 276 packages in 1s

58 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
** Invoke vite:build_all (first_time)
** Invoke vite:verify_install (first_time)
** Execute vite:verify_install
** Execute vite:build_all
Skipping vite build. Watched files have not changed since the last build at 2022-06-19 11:55:23
```

Is there anything that can be done to clobber this warning? 

I guess something somewhere is running using the [deprecated `production` flag](https://docs.npmjs.com/cli/v8/using-npm/config#production). 

```
rake vite:info   
     
bin/vite present?: true
vite_ruby: 3.1.6
vite_rails: 3.0.10
rails: 7.0.3
node: v16.15.1
npm: 8.11.0
yarn: 
pnpm: 
ruby: ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-linux]

installed packages:
myapp@0.1.0 /home/jim/files/Rails/myapp
├─┬ @vitejs/plugin-vue@2.3.3
│ └── vite@2.9.12 deduped
├─┬ vite-plugin-ruby@3.0.12
│ └── vite@2.9.12 deduped
├── vite@2.9.12
└─┬ vitest@0.13.0
  └── vite@2.9.12 deduped
```

Happy to provide more info if required.

**Comment by ElMassimo** [comment]
Hi James! 

The `install_dependencies` task is [using](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/tasks/vite.rake#L47) the [`ci` package](https://github.com/privatenumber/ci) to handle all package installers (`npm`, `pnpm`, and `yarn`).

It doesn't seem to be setting the `--production` flag, so perhaps it's coming from an environment variable in your deployment, such as `NPM_PRODUCTION`.

**Reply by jameshibbard** [reply]
Really? Perhaps I'm misunderstanding the warning.

I had understood npm to be complaining about the use of the deprecated `--production` flag (https://docs.npmjs.com/cli/v8/using-npm/config#production). Would this not be the case?

I get the same error if I leave out the `NODE_ENV` variable. I had included that in my example, as setting that to `production`, should set `--omit=dev`, which I what I thought npm was complaining about (https://docs.npmjs.com/cli/v8/using-npm/config#omit).

I also tried editing the rake task directly (e.g. `system({ 'NODE_ENV' => 'production' }, cmd)`, or `cmd = legacy_flag ? 'npx ci --yes' : 'npx --yes ci --omit=dev'`), but whatever I tried, I was still seeing the warning. And admittedly, by this time I was just shooting in the dark.

> such as `NPM_PRODUCTION`.

I was unaware of this variable. It is not present in my environment, and setting it explicitly to either `true` or `false` had no effect.

Do you have any idea where else I could look, or what else I could try to get rid of this warning?

Thanks again for your help.

**Comment by mattbrictson** [comment, marked as answer]
I am pretty certain this is something `vite_ruby` is doing. You can see here that it is explicitly setting the `NPM_CONFIG_PRODUCTION` env variable. I think that is what leads to the deprecation warning.

https://github.com/ElMassimo/vite_ruby/blob/18cc7299278aba7ce180f7bf8145b7a746cb2c59/vite_ruby/lib/tasks/vite.rake#L85

**Reply by ElMassimo** [reply]
Thanks for finding that Matt! PRs to update this to the latest npm/yarn equivalent are welcome!
